pilot/pkg/config/kube/crdclient/client.go

```diff
@@ -0,0 +1,353 @@
+		return nil
+	}
+	if features.EnableCRDValidation {
+		schema, _ := cl.Schemas().FindByGroupVersionKind(typ)
+		if err = schema.Resource().ValidateProto(cfg.Name, cfg.Namespace, cfg.Spec); err != nil {
+			handleValidationFailure(cfg, err)
+			return nil
+		}
+
+	}
+	return cfg
+}
+
+// Create implements store interface
+func (cl *Client) Create(config model.Config) (string, error) {
+	if config.Spec == nil {
+		return "", fmt.Errorf("nil spec for %v/%v", config.Name, config.Namespace)
+	}
+
+	meta, err := create(cl.istioClient, cl.serviceApisClient, config, getObjectMetadata(config))
+	if err != nil {
+		return "", err
+	}
+	return meta.GetResourceVersion(), nil
+}
+
+// Update implements store interface
+func (cl *Client) Update(config model.Config) (string, error) {
+	if config.Spec == nil {
+		return "", fmt.Errorf("nil spec for %v/%v", config.Name, config.Namespace)
+	}
+
+	meta, err := update(cl.istioClient, cl.serviceApisClient, config, getObjectMetadata(config))
+	if err != nil {
+		return "", err
+	}
+	return meta.GetResourceVersion(), nil
+}
+
+// Delete implements store interface
+func (cl *Client) Delete(typ resource.GroupVersionKind, name, namespace string) error {
+	return delete(cl.istioClient, cl.serviceApisClient, typ, name, namespace)
+}
+
+// List implements store interface
+func (cl *Client) List(kind resource.GroupVersionKind, namespace string) ([]model.Config, error) {
+	h, f := cl.kinds[kind]
+	if !f {
+		return nil, fmt.Errorf("unrecognized type: %s", kind)
+	}
+
+	list, err := h.lister(namespace).List(klabels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	out := make([]model.Config, 0, len(list))
+	for _, item := range list {
+		cfg := TranslateObject(item, kind, cl.domainSuffix)
+		if features.EnableCRDValidation {
+			schema, _ := cl.Schemas().FindByGroupVersionKind(kind)
+			if err = schema.Resource().ValidateProto(cfg.Name, cfg.Namespace, cfg.Spec); err != nil {
+				handleValidationFailure(cfg, err)
+				// DO NOT RETURN ERROR: if a single object is bad, it'll be ignored (with a log message), but
+				// the rest should still be processed.
+				continue
+			}
+
+		}
+
+		if cl.objectInRevision(cfg) {
+			out = append(out, *cfg)
+		}
+	}
+
+	return out, err
+}
+
+func (cl *Client) objectInRevision(o *model.Config) bool {
+	configEnv, f := o.Labels[label.IstioRev]
+	if !f {
+		// This is a global object, and always included
+		return true
+	}
+	// Otherwise, only return if the
+	return configEnv == cl.revision
+}
+
+// knownCRDs returns all CRDs present in the cluster, with retries
+func knownCRDs(crdClient apiextensionsclient.Interface) map[string]struct{} {
+	delay := time.Second
+	maxDelay := time.Minute
+	var res *v1beta1.CustomResourceDefinitionList
+	for {
+		var err error
+		res, err = crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().List(context.TODO(), metav1.ListOptions{})
+		if err == nil {
+			break
+		}
+		log.Errorf("failed to list CRDs: %v", err)
+		time.Sleep(delay)
+		delay *= 2
```

Comment:
How did we get this number for delay? I see a lot of places use it though.

Reply:
it's just standard exponential back off. This is copied from the current code btw, just moved. Basically if this is happening it means either the API server is down or we somehow don't have rbac permissions that we require so no need to aggressively retry and we cap out at 1 min retry interval